File: typescript/bob/bob.ts
class Bob {

    hey(statement: string): string {
        const trimmed = statement.trim()
        if (this.isShouting(trimmed)) {
            return "Whoa, chill out!"
        }
        else if (this.isQuestion(trimmed)) {
            return "Sure."
        }
        else if (this.isSilence(trimmed)) {
            return "Fine. Be that way!"
        }
        else {
            return "Whatever."
        }
    }

    private isShouting(statement: string): boolean {
        const letters = statement.match("[A-Za-zäöüß]+")
        const uppercase = letters !== null ? letters.filter((x) => x === x.toUpperCase()) : []
        return letters !== null && letters.length > 0 && uppercase !== null && letters.length === uppercase.length
    }

    private isSilence(statement: string): boolean {
        return !statement || statement.length === 0
    }

    private isQuestion(statement: string): boolean {
        return statement.endsWith("?")
    }
}

export default Bob

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = {
    def isShouting(statement: String): Boolean = {
      var letters = statement.filter(x => x.isLetter);
      
      !letters.isEmpty() && letters.forall(x => x.isUpper);
    };
    
    def isQuestion(statement: String): Boolean  = {
      statement.endsWith("?");
    };
    
    def isSilence(statement: String): Boolean  = {
      statement.isEmpty();
    };
    
    val trimmed = statement.trim();

    if(isSilence(trimmed))
    {
      "Fine. Be that way!"
    }
    else if(isShouting(trimmed))
    {
      "Whoa, chill out!"
    }
    else if(isQuestion(trimmed))
    {
      "Sure."
    }
    else
    {
      "Whatever."
    }
  }
}
